### DISCUSSION TITLE
smart meter interface with modbus and MAX485 wiring issue

### DISCUSSION THREAD

**Discussion by riker65** [arendst/Tasmota#10795] [Q&A] [answered]
Hi

I am confused and need some advise/confirmation about the wiring.

smart meter interface with modbus and MAX485
[https://tasmota.github.io/docs/Smart-Meter-Interface/](url)

`ESP| MAX485
TXD ->DI
d3 -> DE
d2 ->  RE
RXD -> RO
`
Some use DE and RE and connect it to same GPIO

[https://tasmota.github.io/docs/Smart-Meter-Interface/#modbus-devices](url)
and [https://tasmota.github.io/docs/Smart-Meter-Interface/#meter-declaration](url)

Please confirm the correct wiring.

**Meter Declaration~
+M,rxGPIO,type,flag,parameter,jsonPrefix,txGPIO,txPeriod,cmdTelegram**

so is : rxGPIO -> is DE on MAX485 and txGPIO -> RE? 

`>D

>B
=>sensor53 r

>M 1

+1,14,m,1,9600,EB,5,50,0104006C,01040079,0104007A,0104007F,01040026,01040027,01040028,0104000B,01040084
`

How to code: Serial 8N2? 

Thanks a lot

**Comment by barbudor** [comment, marked as answer]
On a MAX485, RE is Receiver Driver Enable (inverted) and DE is Driver Output Enable
Refer to the below diagram in the [MAX485 datasheet](https://datasheets.maximintegrated.com/en/ds/MAX1487-MAX491.pdf)

![image](https://user-images.githubusercontent.com/2996042/106579406-f3f27e80-6540-11eb-8c3e-b5f327787198.png)

It is the receiver pin RO that should be connected to the GPIO that you specify as rxGPIO in the descriptor.
And the transmitter input pin DI that should be connected to the GPIO specified as txGPIO in the descriptor.

Regarding RE and DE, these are Enable pins.
Originally RS485 (on which MODBUS is based) is a multi-device bidirectional bus which means that at a given time, only 1 device is allowed to speak while the others are listening.
Having RE=LOW enables the receiver, HIGH disables the receiver
Having DE=HIGH enables the transmitter, LOW disables the transmitters
On most system, the 2 pins are connected together and are tied to a GPIO which allow a RS485 device to control if it is speaking (HIGH) or listening (LOW).
Unfortunately Tasmota MODBUS implementation doesn't seem to support driving those pins.
There should be no problem to connect the RE pin to GND to have it continuously LOW and receiver enabled.
But the transmitter should be enabled ONLY when the ESP has to speak. Otherwise it will prevent the meter to speak.

In this section https://tasmota.github.io/docs/Smart-Meter-Interface/#4-hiking-dds238-2-zns3-modbus, the diagram is using another board to do the MODBUS to serial conversion. IT seems to be this board : https://www.aliexpress.com/item/1005001621746811.html
I don't know this board for sure but I would expect that it includes another microcontroller which manage the DE pin.
I would strongly suggest that you get such board to connect an ESP to a true RS485 MODBUS.
In that case, only rxGPIO and txGPIO would be required.

Otherwise, some code changes would be required to introduce a new type of GPIO to control the RE/DE pins.

Note that some meters are referring to MODBUS for the logical protocol (framing, CRC) but are still using standard serial "RS232" style instead of RS485 bi-directional bus. In that case such RS485 module are not required. Only RS23 to TTL(3V) adapter would be required.

Hope this helps

**Comment by riker65** [comment]
Hi Barbodur,

thanks a lot for your explanation.

I am actually using this board:   : 

> https://www.aliexpress.com/item/1005001621746811.html


As I understood  I have to map and wire:

Ttl-RS485	ESP	SMI-Meter Declaration

DI	d0-TX	txGPIO
DE	NA          NA
RE	NA     	NA
RO	D1-RX	rxGPIO

Will test again.

**Reply by barbudor** [reply]
Cool, let us know

By the way, as you have the board, could you be kind enough to let me know the marking on the 2 chips (although I assume the small one is a MAX485 or chinese clone).
If you could also provide a good photo of the component side so I can try to follow the PCB tracks and guess the schematic that would be super kind.
Thanks

**Comment by riker65** [comment]
sorry, I was wrong,


I have the max485 one

[(https://de.aliexpress.com/item/32513892030.html)]

just ordered the others:-)


btw.

do you know how to dump  the sensor53?

I tried

=> sensor53 d1


but do not see any events

**Comment by riker65** [comment]
Hi Barbudor,

having still a lack of info or understanding 

looking into the wiring  I am wondering where to define the D1 pin in the script.

https://i.stack.imgur.com/fLIWw.png


> On most system, the 2 pins are connected together and are tied to a GPIO which allow a RS485 device to control if it is speaking (HIGH) or listening (LOW).
Unfortunately Tasmota MODBUS implementation doesn't seem to support driving those pins.
There should be no problem to connect the RE pin to GND to have it continuously LOW and receiver enabled.

Probably need this to get this working.

**Comment by barbudor** [comment]
As I explained , Tasmota has not implemented the control of pins DE and RE of the MAX485. So there is no software feature to control what is connected to D1 on this schematic. I don't know where you got that diagram. It is surely perfectly exact to work with a software that has implemented such control.
I not sure what is done on the board you have ordered but I found some schematics on the Internet that connects DE to RE to DI all together. Not what I would qualify as a perfect implementation but may work.
One question : do you need to send messages to your meter ? 
If not and you are interested only in receive, just connect DE + RE to GND, and you don't even need to define txGPIO.

I'm not sure what you meant by "dumping sensor53". Could you please elaborate ?

**Comment by riker65** [comment]
Hi 
https://tasmota.github.io/docs/Commands/#sensor53

Sensor53 | 
there is stated:

> d<x> = disable data decoding and dump meter (x = 1..5) data to the Console. This is used to decipher the meter's data format to define the variable encoding in the meter's descriptor.

So I wanted to snif more or less the data.

Do not need sending I think, not at the beginning.
I think I have to send to get certain results delivery.

MY Device is also using Serial 8N2 , not sure if this is supported by Tasmota.

used M in the meter declaration

**Comment by riker65** [comment]
Hi 
checked, I need to send data to get certain answers from the device.
To I need another RS485 ttl adapter.

Do you know which would work with tasmota ? Thanks

**Reply by barbudor** [reply]
connecting RE + DE to DI and TXD won't work sorry
I think you need the board I suggested

**Reply by riker65** [reply]
https://de.aliexpress.com/item/1005001621746811.html

that was the one. Ordered. 

Will see, thanks a lot

**Comment by kugelkopf123** [comment]
This one https://www.amazon.de/DSD-TECH-SH-U12-MAX13487-Raspberry/dp/B07B667STP

But I would suggest to connect your de and re over 10k to bcc as seen in this picture. ![image](https://user-images.githubusercontent.com/45996965/106686548-c8b47180-65ca-11eb-9645-17336e191973.jpeg)

Sensor53 d1 only delivers data if you receive data from your modbus device. 

As explained on the sml docs you have to define a tx and rx pin, poll time and the registers you are asking for. Have a look on the sml page there are some examples of some modbus devices. Look into your data sheet of the modbus device and search for the registers. 
You can compare the use of the registers with the data sheet from one of the modbus examples. 
I be made a tutorial in German language for that. So.. if you are German this could probably help you.

**Reply by barbudor** [reply]
@kugelkopf123 This board would work only if you have something to drive RE and DE when needed
If you just leave DE and RE with pullups and not connected to a GPIO, then Receiver will never be enabled and no reception will occur
On MAX485 one MUST drive RE+DE to state the flow direction : RE=DE=High when listening, RE=DE=Low when transmitting.
As of today Tasmota have not implement any control for the flow direction on RS485

**Reply by riker65** [reply]
Hi @kugelkopf123,

please provide the link  to your tutorial I am german:-) thx

**Comment by gemu2015** [comment]
@riker65 

Your board should work. As far as i can see it has A,B,GND on the Modus side and GND,RX,TX,VCC on the esp side
You need a board that automatically switches the duplex serial line to transmit when a send is detected and immediately returns to receive after a small delay.
Thats exactly what this board should do. (the max485 alone is not supported on Tasmota)

if you connect the esp side correctly you should see the transmit led blink in regular intervals.

if your MODBUS device responds you should also see the receive led blink.
if it does not, your baudrate or serial setting is not correct.
you mentioned your meter has 8N2.
you have to specify this by a 2 after the mode type (m2)

+1,14,m2,1,9600,EB,5,50,0104006C,01040079,0104007A,0104007F,01040026,01040027,01040028,0104000B,01040084

**Reply by barbudor** [reply]
Just to add that Gemu's comment applies to the 2nd board you have just ordered, not the 1st one

**Reply by riker65** [reply]
thanks for clarification and confirmation

**Comment by barbudor** [comment]
Hi @riker65 Is your meter a SDM120 ?
I made a change in SDM120 driver to drive the direction pins of a MAX485 but don't have the device myself to test

HI @gemu2015 
I made a change in TasmotaModbus in order to support direction pin. May be it could be interesting to support this also in script's smart-meter ? However I saw that the script smart-meter interface is based directly on TasmotaSerial and not on TasmotaModbus.

**Reply by riker65** [reply]
Hi Barbudor, 
ok perfect thanks. How is the direction be controlled? 
what do I have to do? within the script any special setting for the control pin?

thanks

**Reply by barbudor** [reply]
@riker65 just to be clear, I've done the change in the SDM120 driver only, not in smart-meter scripting.
So you could test only if you have a SDM120.
As you didn't stated earlier which meter you have... Is that the case?

**Reply by riker65** [reply]
Hi Barbudor,
ok thanks. sorry for maybe caused confusion. I only hat PIZM017 - not tested - and the RS485 MAX - using with the script.

**Reply by iosifnicolae2** [reply]
hi @barbudor 
Could you please share your code so I can start implementing the direction pins? (I'm kind of not willing to wait 2 months for the RS485 to TTL converter to arrive 😊  )

**Reply by barbudor** [reply]
I dropped the code as the PR was not accepted

**Comment by antonwantstosleep** [comment]
Hi @barbudor, @riker65, is there any news to the wiring problem?

I`ve got a russian Neva Energy Meter with RS485 interface and iec-61107 protocol.

Means I need to:
- send 2F 3F 21 0D 0A to start the communication
- receive identity (ex. 2F 54 50 43 35 4E 45 56 41 4D 54 33 32 34 2E 32 35 30 37 0D 0A)
- send 06 30 35 31 0D 0A to set read mode
- get ack and all current active energy values
- send 01 52 31 02 30 30 30 39 30 32 46 46 28 29 03 68 (this is an OBIS code 00.09.02.FF to read 'date')
           SOH R 1  STX 0 0  0  9  0  2  F  F  (  )  ETX BCC
- receive the date
- and so on

I`ve got a ESP8266 and ESP32... and a  MAX485 from ali with DI, DE, RE, RO. Means I need a GPIO to enable / disable transmission.

I need to buy another RS485 to TTL chip, right?

**Reply by Jason2866** [reply]
Only this one is supported https://github.com/jarcikw/SDM120_Homeassistant/wiki/SDM120-&-nodemcu-&-RS485-wiring

**Reply by antonwantstosleep** [reply]
@Jason2866, you talk about that RS485 chip? Can you throw me a link, maybe to ali?

**Reply by barbudor** [reply]
€ 2,17 | 5pcs TTL turn RS485 module 485 to serial UART level mutual conversion hardware automatic flow control
https://a.aliexpress.com/_vXmhBC

**Comment by antonwantstosleep** [comment]
@barbudor, another question?
I managed to get data from the russian TAIPIT energy meter with this script:
```
>D
>B
=>sensor53 r
>M 1

;Get discovery,set read mode
+1,3,o,0,9600,meter01,1,10,2F3F210D0A,063035300D0A

;The responce is 0F0880FF(010674.34,006784.54,003889.80,000000.00,000000.00)!
;  1 - Active energy total KWh
;  2...5 - Active energy tariff 1...4 KWh 
1,0F0880FF(@1,Active Energy Total,KWh,energy_active_total,2

#
```
1. As you can see the standart responce is OBIS-like but different from the docs... I don't understand how to get values from that comma separated string. Now it's getting only FIRST value (strange but it's working).

2. When I send 2F3F210D0A and 063035300D0A to the meter I always get ONLY that responce and nothing more. To get additional values (like current, voltage etc.) I need a more deep conversation with the meter:
- reguest model with 01523102363030313034464628290360 or SOH R1 STX 60 01 04 FF (  ) ETX BCC
- get responce 023630303130344646283030303030303038290309 or STX 60 01 04 FF (00000008) ETX BCC
- and so on
Should I use additional scripting with sml()? Or maybe I should switch to Berry scripting (my board is ESP32)?

**Reply by barbudor** [reply]
I'm not an expert in Scripting and SML. May be @gemu2015 can help 

Alternatively, Berry now support Serial (I've not yet tested but willing to) but AFAIK, there is no interface with Tasmota Energy. @s-hadinger ?

**Comment by gemu2015** [comment]
this is not standard obis.
multiple obis values should look like this and are supported
0-0:98.1.0(210201000000W)(000000.000*kWh)(000000.000*kWh))

there is currently no support for this format 
you may try with berry

**Comment by iosifnicolae2** [comment]
Hey, I've implemented a solution here: https://github.com/arendst/Tasmota/pull/17247

**Comment by grzesiekn** [comment]
Make sure you buy the 3.3V version - the MAX485 chip requires 5V so adapters with the MAX3485 are needed, most sellers on aliexpress sell the MAX485 adapters as 3.3V/5V and I haven't been able to get any of these to work with ESP.
I ended up buying a MAX3845 IC and soldering it to the modules instead of the MAX485.

On aliexpress there are modules with MAX3485 (they don't have the TTL inscription on the back), but in 2 cases I got modules with the TTL inscription and the MAX485 chip (other than in the pictures).
